kpi/permissions.py

```diff
@@ -52,3 +56,49 @@ class PostMappedToChangePermission(IsOwnerOrReadOnly):
     '''
     perms_map = IsOwnerOrReadOnly.perms_map
     perms_map['POST'] = ['%(app_label)s.change_%(model_name)s']
+
+
+class AssetOwnerNestedObjectsPermissions(permissions.BasePermission):
+    """
+    Permissions for objects that are nested under Asset which only owner should access.
+    Others should receive a 404 response (instead of 403) to avoid to reveal existence
+    of objects.
+    """
+    def has_permission(self, request, view):
+
+        if not request.user or (request.user and
+                                (request.user.is_anonymous() or
+                                not request.user.is_authenticated())):
+            return False
+
+        asset_uid = self.__get_parents_query_dict(request).get("asset")
+        asset = get_object_or_404(Asset, uid=asset_uid)
+
+        if request.user != asset.owner:
+            raise Http404
+
+        return True
+
+    def has_object_permission(self, request, view, obj):
```

Comment:
would `obj` here be useful? i was wondering if we could we say `obj.asset` here instead of `__get_parents_query_dict(request).get("asset")` etc. in `has_permission()`.

Reply:
Actually, no. 
`obj` is only available when we are in the `detail` routes. e.g.:
- `/hook/<hook_uid>/`
- `/hook/<hook_uid>/logs/<hook_log_uid>`

It's not available in `list`routes. e.g.:
- `/hook/`
- `/hook/<hook_uid>/logs/`
- `/versions/`

Moreover,  `HookLog` objects don't have any relationships with `Asset`, so `obj.asset` would fail.